src/dag.c

```diff
@@ -901,399 +901,447 @@ void RedisAI_Disconnected(RedisModuleCtx *ctx, RedisModuleBlockedClient *bc) {
+				rinfo->dagDeviceOps = array_append(rinfo->dagDeviceOps, rinfo->dagOps[j]);
+			}
+		}
+
+		rinfo->dagDeviceOpCount = array_len(rinfo->dagDeviceOps);
+	}
+
+	for (long long i=0; i<ndevices; i++) {
+		const char* devicestr = devices[i];
+		RunQueueInfo *run_queue_info = NULL;
+		if (ensureRunQueue(devicestr, &run_queue_info) == REDISMODULE_ERR) {
+			RAI_FreeRunInfo(rinfo);
+			return REDISMODULE_ERR;
+		}
+
+		RedisAI_RunInfo *rinfo = rinfo_copies[i];
+
+		gettimeofday(&rinfo->queuingTime, NULL);
+
+		pthread_mutex_lock(&run_queue_info->run_queue_mutex);
+		queuePush(run_queue_info->run_queue, rinfo);
+		pthread_cond_signal(&run_queue_info->queue_condition_var);
+		pthread_mutex_unlock(&run_queue_info->run_queue_mutex);
+	}
+	array_free(devices);
+	array_free(rinfo_copies);
+
+	return REDISMODULE_OK;
+}
 
-    rinfo->dagDeviceOpCount = array_len(rinfo->dagDeviceOps);
-  }
 
-  for (long long i=0; i<ndevices; i++) {
-    const char* devicestr = devices[i];
-    RunQueueInfo *run_queue_info = NULL;
-    if (ensureRunQueue(devicestr, &run_queue_info) == REDISMODULE_ERR) {
-      RAI_FreeRunInfo(ctx, rinfo);
-      return RedisModule_ReplyWithError(ctx,
-                                        "ERR Queue not initialized for device");
-    }
+void DAG_FinishBlockingExecution (RedisAI_OnFinishCtx ctx, void *private_data) {
 
-    RedisAI_RunInfo *rinfo = rinfo_copies[i];
+	RedisAI_RunInfo *rinfo = (RedisAI_RunInfo *)private_data;
+	if(rinfo->client) RedisModule_UnblockClient(rinfo->client, rinfo);
+}
 
-    gettimeofday(&rinfo->queuingTime, NULL); 
+int RedisAI_DagRunSyntaxParser(RedisModuleCtx *ctx, RedisModuleString **argv,
+                                 int argc, int dagMode) {
 
-    pthread_mutex_lock(&run_queue_info->run_queue_mutex);
-    queuePush(run_queue_info->run_queue, rinfo);
-    pthread_cond_signal(&run_queue_info->queue_condition_var);
-    pthread_mutex_unlock(&run_queue_info->run_queue_mutex);
-  }
+	RedisAI_RunInfo *rinfo = NULL;
+	if (RAI_InitRunInfo(&rinfo) == REDISMODULE_ERR) {
+		return RedisModule_ReplyWithError(ctx,
+		  "ERR Unable to allocate the memory and initialise the RedisAI_RunInfo structure");
+	}
+	// Parse DAG string command and store the data in rinfo obj.
+	int curr_status = DAG_StringParser(ctx, argv, argc, dagMode, &rinfo);
+	if(curr_status != REDISMODULE_OK) return REDISMODULE_ERR;
+	// Block the client before adding rinfo to the run queues (sync call)
+	rinfo->client = RedisModule_BlockClient(ctx, RedisAI_DagRun_Reply, NULL, RedisAI_FreeData, 0);
+	RedisModule_SetDisconnectCallback(rinfo->client, RedisAI_Disconnected);
+	// Use the entire rinfo obj as the on finish's private data.
+	rinfo->private_data = rinfo;
+	rinfo->OnFinish = DAG_FinishBlockingExecution;
+	curr_status = DAG_InsertDAGToQueue(rinfo);
+	if(curr_status == REDISMODULE_ERR) {
+		RedisModule_AbortBlock(rinfo->client);
+		return RedisModule_ReplyWithError(ctx, "ERR Queue not initialized for device");
```

Comment:
move to a function that should be shared between HLAPI and LLAPI
and set a global error in run info